erts/emulator/beam/erl_db_util.c

```diff
@@ -2804,6 +2805,72 @@ Eterm db_prog_match(Process *c_p,
                }
             }
             break;
+        case matchCurrentStacktrace: {
+            Uint sz;
+            Uint heap_size;
+            Eterm *hp;
+            Eterm mfa;
+            struct StackTrace *s;
+            int depth;
+            FunctionInfo* stk;
+            FunctionInfo* stkp;
+
+            ASSERT(c_p == self);
+
+            depth = unsigned_val(esp[-1]);
+
+            if (depth > erts_backtrace_depth) {
+                depth = erts_backtrace_depth;
+            }
+
+            sz = offsetof(struct StackTrace, trace) + sizeof(ErtsCodePtr) * depth;
+            s = (struct StackTrace *) erts_alloc(ERTS_ALC_T_TMP, sz);
+            s->depth = 0;
+            s->pc = NULL;
+
+            erts_save_stacktrace(c_p, s, depth);
+
+            depth = s->depth;
+            stk = stkp = (FunctionInfo *) erts_alloc(ERTS_ALC_T_TMP,
+					     depth*sizeof(FunctionInfo));
+
+            heap_size = 0;
+            for (i = depth - 1; i >= 0; i--) {
+                erts_lookup_function_info(stkp, s->trace[i], 1);
+                if (stkp->mfa) {
+                    heap_size += stkp->needed + 2;
+                    stkp++;
+                }
+            }
+
+            if (stkp == stk) {
+                *esp++ = NIL;
+
+                erts_free(ERTS_ALC_T_TMP, stk);
+                erts_free(ERTS_ALC_T_TMP, s);
+
+                break;
+            }
+
+            ehp = HAllocX(build_proc, 2 * depth + heap_size, HEAP_XTRA);
+            *esp++ = make_list(ehp);
+            hp = &ehp[2 * depth];
+
+            while (stkp > stk) {
+                stkp--;
+                hp = erts_build_mfa_item(stkp, hp, am_true, &mfa, NIL);
+                *ehp++ = mfa;
+                *ehp = make_list(ehp + 1);
+                ehp++;
+            }
+
+            ehp[-1] = NIL;
```

Comment:
This section is a bit odd: we've already calculated the amount of heap space we need in the loop above this, so we don't need to allocate an extra `2 * depth` elements. It should also be a bit easier to build the list using our macros, e.g.

```suggestion
            Eterm res = NIL; /* declare where appropriate. feel free to
                              * clean up the declarations at the top
                              * while you're at it. */

            if (heap_size > 0) {
                ehp = HAllocX(build_proc, heap_size, HEAP_XTRA);

                ASSERT((stkp - stk) > 0 && (stkp - stk) < MAX_BACKTRACE_SIZE);
                for (i = stkp - stk; i >= 0; i--) {
                    Eterm mfa;

                    ehp = erts_build_mfa_item(stk[i], ehp, am_true, &mfa, NIL);
                    res = CONS(ehp, mfa, res);
                    ehp += 2;
                }
            }

            *esp++ = res;
```

(I haven't tested the code above)

Reply:
Thanks for your review, @jhogberg! :) It is better, I have applied it. However, I am unsure what cleaning up the declarations mean. Please advise.

Reply:
There were a bunch of variable declarations near `case matchCurrentStacktrace: {` that are no longer needed, but I see you've already done that :)